PolyEngine/Core/Src/BTreeNode.hpp

```diff
@@ -0,0 +1,605 @@
+					};
+				};
+				AscensionResult const ascend() {
+					if (node->parent) {
+						return {KVERef{ NodeRef{ height + 1u, node->parent, root }, node->edge_idx} };
+					} else {
+						return {*this};
+					}
+				};
+
+				template<typename Key, typename Val>
+				void push_back(Key&& key, Val&& value) {
+					//leaf
+					ASSERTE(node->len < CAPACITY, "No space left in the leaf!");
+
+					const uint16_t idx = node->len;
+
+					::new(&node->keys  [idx]) K(std::forward<Key>(key));
+					::new(&node->values[idx]) V(std::forward<Val>(value));
+
+					node->len += 1u;
+				}
+
+				template<typename Key, typename Val>
+				void push_front(Key&& key, Val&& value) {
+					//leaf
+					ASSERTE(node->len < CAPACITY, "No space left in the leaf!");
+
+					constructing_move_backwards(node->keys.begin(),   node->keys.begin()   + node->len, node->keys.begin()   + node->len + 1u);
+					::new(&node->keys  [0]) K(std::forward<Key>(key));
+
+					constructing_move_backwards(node->values.begin(), node->values.begin() + node->len, node->values.begin() + node->len + 1u);
+					::new(&node->values[0]) V(std::forward<Val>(value));
+
+					node->len += 1u;
+				}
+
+				template<typename Key, typename Val>
+				void push_back(Key&& key, Val&& value, const Edge edge) {
+					//branch
+					ASSERTE(edge.height == height - 1u, "Attempting to push an edge node of incorrect height!");
+					ASSERTE(node->len < CAPACITY, "No space left in the branch!");
+
+					const size_t idx = node->len;
+
+					::new(&node->keys  [idx]) K(std::forward<Key>(key));
+					::new(&node->values[idx]) V(std::forward<Val>(value));
+					node->as_branch()->edges[idx + 1u] = edge.node;
+
+					node->len += 1u;
+
+					KVERef{*this, idx + 1u}.correct_parent_link();
+				}
+
+				template<typename Key, typename Val>
+				void push_front(Key&& key, Val&& value, const Edge edge) {
+					//branch
+					ASSERTE(edge.height == height - 1u, "Attempting to push an edge node of incorrect height!");
+					ASSERTE(node->len < CAPACITY, "No space left in the branch");
+
+					constructing_move_backwards(node->keys.begin(),   node->keys.begin()   + node->len, node->keys.begin()   + node->len + 1u);
+					::new(&node->keys  [0]) K(std::forward<Key>(key));
+
+					constructing_move_backwards(node->values.begin(), node->values.begin() + node->len, node->values.begin() + node->len + 1u);
+					::new(&node->values[0]) V(std::forward<Val>(value));
+
+					auto& edges = node->as_branch()->edges;
+					std::move_backward(edges.begin(), edges.begin() + node->len + 1u, edges.begin() + node->len + 2u);
+					edges[0] = edge.node;
+
+					node->len += 1u;
+
+					for (size_t i = 0; i < node->len + 1u; i++) {
+						KVERef{*this, i}.correct_parent_link();
+					}
+				}
+
+				struct PopResult {
+					K key;
+					V value;
+					Edge edge; //optional
+				};
+
+				PopResult pop_back() {
+					ASSERTE(node->len > 0, "No entry to pop in the node!");
+
+					const uint16_t idx = node->len - 1u;
```

Comment:
Performance/space overhead. It's quite unlikely you're going to need more than 65535 elements in one node (default is 11 :P) and uint16_t fits in nicely.